src/python/pants/backend/jvm/subsystems/dependency_context.py

```diff
@@ -0,0 +1,150 @@
+# coding=utf-8
+# Copyright 2014 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import hashlib
+
+from pants.backend.codegen.thrift.java.java_thrift_library import JavaThriftLibrary
+from pants.backend.jvm.targets.annotation_processor import AnnotationProcessor
+from pants.backend.jvm.targets.jar_library import JarLibrary
+from pants.backend.jvm.targets.javac_plugin import JavacPlugin
+from pants.backend.jvm.targets.jvm_target import JvmTarget
+from pants.backend.jvm.targets.scalac_plugin import ScalacPlugin
+from pants.base.exceptions import TargetDefinitionException
+from pants.base.fingerprint_strategy import FingerprintStrategy
+from pants.build_graph.address import Address
+from pants.build_graph.aliased_target import AliasTarget
+from pants.build_graph.resources import Resources
+from pants.build_graph.target import Target
+from pants.build_graph.target_scopes import Scopes
+from pants.subsystem.subsystem import Subsystem
+
+
+class SyntheticTargetNotFound(Exception):
+  pass
+
+
+class DependencyContext(Subsystem):
+  """Implements calculating `exports` and exception (compiler-plugin) aware dependencies.
+
+  This is a subsystem because in future the compiler plugin types should be injected
+  via subsystem or option dependencies rather than declared statically.
+  """
+
+  options_scope = 'jvm-dependency-context'
+
+  _target_closure_kwargs = dict(include_scopes=Scopes.JVM_COMPILE_SCOPES, respect_intransitive=True)
+  _compiler_plugin_types = (AnnotationProcessor, JavacPlugin, ScalacPlugin)
+
+  @classmethod
+  def _get_synthetic_target(cls, target, thrift_dep):
+    """Find a thrift target's corresponding synthetic target."""
+    for dep in target.dependencies:
+      if dep != thrift_dep and dep.is_synthetic and dep.derived_from == thrift_dep:
+        return dep
+    return None
+
+  @classmethod
+  def _resolve_strict_dependencies(cls, target):
+    for declared in target.dependencies:
+      if type(declared) in (AliasTarget, Target):
+        # Is an alias. Recurse to expand.
+        for r in cls._resolve_strict_dependencies(declared):
+          yield r
+      else:
+        yield declared
+
+      for export in cls._resolve_exports(declared):
+        yield export
+
+  @classmethod
+  def _resolve_exports(cls, target):
+    for export in getattr(target, 'exports', []):
+      if not isinstance(export, Target):
+        addr = Address.parse(export, relative_to=target.address.spec_path)
+        export = target._build_graph.get_target(addr)
```

Comment:
This is existing code migrated from elsewhere. Will avoid making substantive changes if possible.